File: frontend/src/features/connections/pendencias.ts
import type { ConnectionOut } from "../../lib/api";
import type { components } from "../../lib/api-types";

/**
 * Predicados de pendência de segredo pós-import (spec F6 §3.2-8, decisão A-4, achado
 * F6R-14). Espelha `ottima_core.portability.pendencias.pendencias_da_conexao`
 * (packages/ottima-core/src/ottima_core/portability/pendencias.py) byte a byte na
 * semântica das 3 fórmulas — único lugar de verdade no cliente; nenhuma outra tela
 * reimplementa os predicados. Puro: sem I/O.
 *
 * Ordem fixa de exibição, igual nas duas funções abaixo, para a coluna de pendências
 * (spec §6.3-2) não mudar de forma entre renders: senha, certificado_servidor,
 * certificado_aplicacao.
 */
export type Pendencia = "senha" | "certificado_servidor" | "certificado_aplicacao";

type PendingSecretOut = components["schemas"]["PendingSecretOut"];

/**
 * `appCertExiste: null` significa "não avaliável" — o papel operador não lê
 * `GET /api/certificates/app` (`require_admin`, plano F6b decisão de RBAC), então o
 * terceiro predicado simplesmente não aparece; `null` nunca é tratado como `false`.
 */
export function pendenciasDaConexao(
  conexao: Pick<ConnectionOut, "auth_mode" | "has_password" | "security_policy" | "server_cert_file">,
  appCertExiste: boolean | null,
): Pendencia[] {
  const pendencias: Pendencia[] = [];
  if (conexao.auth_mode === "user_password" && !conexao.has_password) {
    pendencias.push("senha");
  }
  if (conexao.security_policy !== "none" && !conexao.server_cert_file) {
    pendencias.push("certificado_servidor");
  }
  if (
    (conexao.security_policy !== "none" || conexao.auth_mode === "certificate") &&
    appCertExiste === false
  ) {
    pendencias.push("certificado_aplicacao");
  }
  return pendencias;
}

/** Traduz o `PendingSecretOut` da resposta de import (tarefa 2.3/3.2-8) para o mesmo
 *  vocabulário — precisa concordar com `pendenciasDaConexao` para a mesma conexão. */
export function pendenciasDoResumo(p: PendingSecretOut): Pendencia[] {
  const pendencias: Pendencia[] = [];
  if (p.needs_password) pendencias.push("senha");
  if (p.needs_server_certificate) pendencias.push("certificado_servidor");
  if (p.needs_app_certificate) pendencias.push("certificado_aplicacao");
  return pendencias;
}

export const ROTULO_PENDENCIA: Record<Pendencia, string> = {
  senha: "Senha",
  certificado_servidor: "Certificado do servidor",
  certificado_aplicacao: "Certificado de aplicação",
};

/** Texto do `title` por pendência (spec §6.3-3): o efeito exato, não um rótulo
 *  genérico. Códigos entre crases são os de `FailureReason`
 *  (`services/opc-worker/src/ottima_opc_worker/security.py`). */
export const EFEITO_PENDENCIA: Record<Pendencia, string> = {
  senha: "A conexão falhará em `connect_failed` até a senha ser informada novamente.",
  certificado_servidor:
    "A conexão falhará em `cert_missing` até confiar no certificado do servidor.",
  certificado_aplicacao:
    "A conexão falhará em `cert_missing` até gerar o certificado de aplicação da instalação.",
};

File: packages/ottima-core/src/ottima_core/portability/pendencias.py
"""Predicados de pendência de segredo pós-import (spec F6 §3.2-8, decisão A-4, achado
F6R-14). Único lugar de verdade das 3 fórmulas — a tarefa 2.3 (import) e o frontend
(F6b) consomem daqui, nunca reimplementam. Puro: nenhuma função aqui toca banco, Redis
ou disco.
"""

from ottima_core.schemas.projects import PendingSecretOut


def pendencias_da_conexao(
    *,
    connection_name: str,
    auth_mode: str,
    has_password: bool,
    security_policy: str,
    server_cert_file: str | None,
    app_cert_exists: bool,
) -> PendingSecretOut:
    """Os 3 predicados de §3.2-8:

    - `needs_password`: autenticação usuário/senha sem senha armazenada.
    - `needs_server_certificate`: policy segura sem o certificado do servidor confiado
      (`server_cert_file` ausente ou vazio).
    - `needs_app_certificate`: policy segura OU autenticação por certificado, sem o par
      de certificado de aplicação da instalação (F6R-14: sem este terceiro predicado,
      uma conexão `auth_mode: certificate` com `security_policy: none` teria pendência
      vazia e falharia em `cert_missing` sem aviso nenhum).
    """
    return PendingSecretOut(
        connection_name=connection_name,
        needs_password=auth_mode == "user_password" and not has_password,
        needs_server_certificate=security_policy != "none" and not server_cert_file,
        needs_app_certificate=(security_policy != "none" or auth_mode == "certificate")
        and not app_cert_exists,
    )
